src/js/components/OptionalPortsAndServiceDiscoveryComponent.jsx

```diff
@@ -0,0 +1,274 @@
+      ? "partially "
+      : "";
+
+    var dynamicPortText = !rows.every(row => !row.isRandomPort)
+      ? ` which will be ${partialRandomText}assigned dynamically`
+      : "";
+
+    var type = determinePortDefinitionsType(this.props.fields);
+
+    var portIdentifiers = Array(rows.length)
+      .fill()
+      .map((_, i) => "$PORT" + i)
+      .join(", ")
+      .replace(/(.*), (.*)$/, "$1 and $2");
+
+    var message = "Your application will need to be configured to listen to" +
+      ` ${portIdentifiers}${dynamicPortText}.`;
+
+    if (type === ContainerConstants.NETWORK.HOST) {
+      message = "Your Docker container will need to be configured to listen" +
+        ` to ${portIdentifiers}${dynamicPortText}.`;
+    } else if (type === ContainerConstants.NETWORK.BRIDGE) {
+      message = "Your Docker container will bind to the requested ports and" +
+        ` they will be dynamically mapped to ${portIdentifiers} on the host.`;
+    }
+
+    return <div>{message}</div>;
+  },
+
+  getRandomPortCheckbox: function (row, i, hidden = false) {
+    var classSet = classNames({
+      "hidden": hidden,
+    }, "checkbox-form-group port-input-field");
+
+    var value = hidden
+      ? false
+      : row.isRandomPort;
+
+    return (
+      <FormGroupComponent className={classSet}
+          fieldId={`${fieldsetId}.${i}.isRandomPort`}
+          label="Assign a random port"
+          value={value}>
+        <input ref={`isRandomPort${i}`} type="checkbox" />
+      </FormGroupComponent>
+    );
+  },
+
+  getPortInputField: function (row, i) {
+    var type = determinePortDefinitionsType(this.props.fields);
+    var isBridgeNetwork = type === ContainerConstants.NETWORK.BRIDGE;
+
+    var fieldLabel = isBridgeNetwork
+      ? "Container Port"
+      : "Port";
+
+    var randomPortField = null;
+
+    if (!isBridgeNetwork && row.isRandomPort) {
+      randomPortField = (
+        <FormGroupComponent
+            label={fieldLabel}
+            value={"$PORT" + i}>
+          <input disabled={true} />
+        </FormGroupComponent>
+      );
+    }
+
+    let portFieldClassSet = classNames({
+      "hidden": !!randomPortField
+    });
+
+    return (
+      <div className="col-sm-4">
+        <FormGroupComponent className={portFieldClassSet}
+            fieldId={`${fieldsetId}.${i}.port`}
+            label={fieldLabel}
+            value={row.port}>
+          <input ref={`port${i}`} {...PortInputAttributes} />
+        </FormGroupComponent>
+        {randomPortField}
+        {this.getRandomPortCheckbox(row, i, isBridgeNetwork)}
+      </div>
+    );
+  },
+
+  getPortDefinitionRow: function (row, i, disableRemoveButton = false) {
+    var error = this.getError(fieldsetId, row.consecutiveKey);
+    var getErrorMessage = this.props.getErrorMessage;
+
+    var rowClassSet = classNames({
+      "has-error": !!error,
+      "duplicable-row": true
+    });
+
+    return (
+      <div key={row.consecutiveKey+i} className={rowClassSet}>
```

Comment:
Good point, but I think it is harmless. Or will the key stay as a number? I think it's converted to a string, isn't it?
No error yet seen, we also use this way at the other panels. :) Think it's ok.